src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/HttpResponseBodyReader.cs

```diff
@@ -54,90 +51,186 @@ public ValueTask<string> ReadAsync(HttpResponseMessage response, CancellationTok
             return new(Constants.UnreadableContent);
         }
 
-        return ReadFromStreamWithTimeoutAsync(response, ResponseReadTimeout, _responseReadLimit, _streamManager,
-            cancellationToken).Preserve();
+        return ReadFromStreamWithTimeoutAsync(response, ResponseReadTimeout, _responseReadLimit, cancellationToken).Preserve();
+    }
+
+    private static async ValueTask<string> ReadFromStreamWithTimeoutAsync(HttpResponseMessage response, TimeSpan readTimeout, int readSizeLimit, CancellationToken cancellationToken)
+    {
+        using var joinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        joinedTokenSource.CancelAfter(readTimeout);
+
+        // TimeSpan.Zero cannot be set from user's code as
+        // validation prevents values less than one millisecond
+        // However, this is useful during unit tests
+        if (readTimeout <= TimeSpan.Zero)
+        {
+            // cancel immediately, async cancel not required in tests
+#pragma warning disable CA1849 // Call async methods when in an async method
+            joinedTokenSource.Cancel();
+#pragma warning restore CA1849 // Call async methods when in an async method
+        }
+
+        try
+        {
+            return await ReadFromStreamAsync(response, readSizeLimit, joinedTokenSource.Token).ConfigureAwait(false);
+        }
+
+        // when readTimeout occurred: joined token source is cancelled and cancellationToken is not
+        catch (OperationCanceledException) when (joinedTokenSource.IsCancellationRequested && !cancellationToken.IsCancellationRequested)
+        {
+            return Constants.ReadCancelled;
+        }
     }
 
-    private static async ValueTask<string> ReadFromStreamAsync(HttpResponseMessage response, int readSizeLimit,
-        RecyclableMemoryStreamManager streamManager, CancellationToken cancellationToken)
+    private static async ValueTask<string> ReadFromStreamAsync(HttpResponseMessage response, int readSizeLimit, CancellationToken cancellationToken)
     {
-#if NET5_0_OR_GREATER
+#if NET6_0_OR_GREATER
         var streamToReadFrom = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
 #else
         var streamToReadFrom = await response.Content.ReadAsStreamAsync().WaitAsync(cancellationToken).ConfigureAwait(false);
 #endif
 
-        var bufferWriter = _bufferWriterPool.Get();
-        var memory = bufferWriter.GetMemory(readSizeLimit).Slice(0, readSizeLimit);
-#if !NETCOREAPP3_1_OR_GREATER
-        byte[] buffer = memory.ToArray();
-#endif
-        try
+        var pipe = new Pipe();
+
+        var bufferedString = await BufferStreamAndWriteToPipeAsync(streamToReadFrom, pipe.Writer, readSizeLimit, cancellationToken).ConfigureAwait(false);
```

Comment:
[nit] No var here please, the type is not obvious

Reply:
Done, thanks